dag/dag_test.go

```diff
@@ -0,0 +1,497 @@
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAddNode(t *testing.T) {
+
+	family := New()
+
+	tests := []struct {
+		name     string
+		nodeId   string
+		nodeName string
+		expErr   error
+	}{
+		{
+			"Add Relation Valid Node",
+			"node-1-id",
+			"node-1-name",
+			nil,
+		},
+		{
+			"Add Relation Valid Node 2",
+			"node-2-id",
+			"node-2-name",
+			nil,
+		},
+		{
+			"Add Relation Invalid Node 2 (node already exists)",
+			"node-2-id",
+			"node-2-name",
+			fmt.Errorf(NodeAlreadyExistsError),
+		},
+	}
+
+	for _, tc := range tests {
+		actualErr := family.AddNode(tc.nodeId, tc.nodeName)
+
+		if tc.expErr == nil {
+			assert.Equal(t, actualErr, tc.expErr, tc.name)
+			assert.NotNil(t, family.nodes[tc.nodeId], tc.name)
+			assert.Equal(t, family.nodes[tc.nodeId].GetId(), tc.nodeId, tc.name)
+			assert.Equal(t, family.nodes[tc.nodeId].GetName(), tc.nodeName, tc.name)
+		} else {
+			assert.ErrorContainsf(t, actualErr, tc.expErr.Error(), tc.name)
+		}
+	}
+}
+
+func TestGetParents(t *testing.T) {
+
+	family := New()
+
+	node1id, node1name := "N1-id", "N1-name"
+	node2id, node2name := "N2-id", "N2-name"
+	node3id, node3name := "N3-id", "N3-name"
+	node4id, node4name := "N4-id", "N4-name"
+	node5id, node5name := "N5-id", "N5-name"
+
+	family.AddNode(node1id, node1name)
+	family.AddNode(node2id, node2name)
+	family.AddNode(node3id, node3name)
+	family.AddNode(node4id, node4name)
+	family.AddNode(node5id, node5name)
+
+	// 4    5
+	//   \ /
+	//    2    3
+	//      \ /
+	//       1
+
+	family.AddRelation(node2id, node1id)
+	family.AddRelation(node3id, node1id)
+	family.AddRelation(node4id, node2id)
+	family.AddRelation(node5id, node2id)
+
+	tests := []struct {
+		name    string
+		nodeId  string
+		parents map[string]*Node
+		expErr  error
+	}{
+		{
+			"Get Parents Valid Node",
+			node1id,
+			map[string]*Node{
+				node2id: family.nodes[node2id],
+				node3id: family.nodes[node3id],
+			},
+			nil,
+		},
+		{
+			"Get Parents valid Node - (root node)",
+			node5id,
+			map[string]*Node{},
+			nil,
+		},
+		{
+			"Get Parents Invalid Node",
+			"node9id",
+			map[string]*Node{},
+			fmt.Errorf(NodeNotFound),
+		},
+	}
+
+	for _, tc := range tests {
+		actualParents, actualErr := family.GetParents(tc.nodeId)
+
+		if tc.expErr == nil {
+			assert.Equal(t, actualErr, tc.expErr, tc.name)
+			assert.Equal(t, actualParents, tc.parents, tc.name)
+		} else {
+			assert.ErrorContainsf(t, actualErr, tc.expErr.Error(), tc.name)
+		}
+	}
+}
+
+func TestGetChildren(t *testing.T) {
+
+	family := New()
+
+	node1id, node1name := "N1-id", "N1-name"
+	node2id, node2name := "N2-id", "N2-name"
+	node3id, node3name := "N3-id", "N3-name"
+	node4id, node4name := "N4-id", "N4-name"
+	node5id, node5name := "N5-id", "N5-name"
+
+	family.AddNode(node1id, node1name)
+	family.AddNode(node2id, node2name)
+	family.AddNode(node3id, node3name)
+	family.AddNode(node4id, node4name)
+	family.AddNode(node5id, node5name)
+
+	// 4    5
```

Comment:
where to put comments?

Reply:
did not find example of writing comment in unit test